### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by ffMathy** [sergeyzwezdin/Hangfire.Mongo#272]
I have 1.5 million scheduled messages, and it's performing quite bad. I know 1.5 million is a lot, but it's constantly making a regex "starts with" query in Mongo to get scheduled messages, which is not very optimized.

Is it possible to make a different query that uses a direct match instead?

**Comment by gottscj** [comment]
@ffMathy,

Thank you for your inquiry.
The challenge is that in order to make sure we dont create duplicates, we need a unique key for the meta data used by Hangfire.Core. we cannot control the keys used by Hangfire.Core and we dont necessarily have transactions available
The compromise is that we are creating a compound key to meet the requirements.
We do have a index on the fields and the regex should take advantage of those.

You can override the default behavior by overriding the "CreateSetFilter" by:
```csharp
public class MyWriteOnlyTransaction : MongoWriteOnlyTransaction
{
    public MyWriteOnlyTransaction(HangfireDbContext dbContext) : base(dbContext)
    {
    }

    public override BsonDocument CreateSetFilter(string key)
    {
        return base.CreateSetFilter(key);
    }
}

public class MyFactory : MongoFactory
{
    public override MongoWriteOnlyTransaction CreateMongoWriteOnlyTransaction(HangfireDbContext dbContext)
    {
        return new MyWriteOnlyTransaction(dbContext);
    }
}
```

You need to use the factory in the MongoStorageOptions like so:
```csharp
var options = new MongoStorageOptions
{
    Factory = new MyFactory()
};
```

let me know if it works out :)

**Comment by ffMathy** [comment]
Thank you for the elaboration. While the regex does indeed get hit by an index, it is far from as efficient as a direct match index.

Instead of having "recurring-job:Foo" as key (or whatever format we use here), can we string split it into two properties with values "recurring-job" and "foo" and match on those?

**Comment by gottscj** [comment]
@ffMathy,

We could probably figure something out.

Could you please try to run the offending queries in mongodb compass eg. and check explain, it would be great if you could post the json result here. I suspect it might the sorting.

`
{Key:{"$regex": "recurring-jobs"}, _t: "SetDto"}
`

**Comment by gottscj** [comment]
Please try with and without sorting: 
`
{_id: 1}
`

**Comment by ffMathy** [comment]
Well it's not that it isn't hitting the index. It's just that a "starts with" regex index hit is not as efficient as one with two fields.

Thank you so far for your involvement 🙏

**Comment by gottscj** [comment]
@ffMathy,

no worries.  You have good data, so its an interesting use-case. :)

When doing the "startswith" query I also use another indexed field: "_t", so I suspect the query is not that invasive.

When you schedule a job Hangfire is hitting the db a couple of time:
1. acquire a lock for the scheduler
2. insert the job
3. update job state and schedule info

besides all this there is a lot other traffic going on

so scheduling 1.5 million jobs generates a lot of traffic.
Would you be able to create recurrent jobs which poll for data instead? or schedule less jobs and have each job handle multiple items?

**Comment by ffMathy** [comment]
I'll report back with more data tomorrow (Monday).

Our Mongo cluster has quite heavy specs, and is hosted on Mongo Atlas.

It is an NVMe disk with 100.000+ IOPS. I can easily query in collections with 4 million items in, in less than 100 milliseconds. So something is causing this particular query to be slow.

We noticed that disk utilization is not the issue, but CPU utilization.

I believe that a direct index is less CPU intensive than a regex starts with index.

So in other words... The index is indeed helping with disk usage. It's the CPU that is a problem.

For other queries, there is no issue. Not even with larger amounts of data.

**Comment by gottscj** [comment]
@ffMathy,

Atlas is flagging the query in the profiler? If we could narrow down which specific query is slowing the system. Its a bit intrusive to change the schema, but it should be possible

**Comment by ffMathy** [comment]
Inside the "Realtime" tab, I caught one of the queries now.

![image](https://user-images.githubusercontent.com/2824010/111966046-c06eb200-8af6-11eb-882a-103e9d3f01af.png)

Note that it says "Index utilized: No".

Here's the full JSON output for that query:

![image](https://user-images.githubusercontent.com/2824010/111966593-63273080-8af7-11eb-9681-a8f2ecd93b4d.png)

But even if it was utilizing an index, I doubt the regex index would be very efficient.

**Comment by gottscj** [comment]
@ffMathy, 

Thanks for the info,

Could you please verify there is an index on both "Key" and "_t" in the "jobGraph" collection? I would expect both indexes to be utilized

Thanks

**Comment by ffMathy** [comment]
Both indexes exist, and they are the right sort order (ascending). I think it's the regex query that makes it not perform well, and not hit the index properly.

Also note that it's only in Mongo Atlas' realtime view that it says it's not utilizing the index. In Compass, it seems to work fine when I run an explain on it:

![image](https://user-images.githubusercontent.com/2824010/112000597-6c77c380-8b1e-11eb-8ea6-d11a64bbef0c.png)

However, do note that it only seems to be using the `Score` index. Shouldn't it use `_t` too?

Also, have you considered using composite indices? I think they can really speed up things, but I am probably not as knowledgeable about Mongo as you are. 

If you decide to go for a composite index, perhaps having `_t` before `Score` in the composite index, if `_t` is more specific. That has a performance difference too.

As a side-node, I noticed this morning that the amount of documents in the database had actually climbed to 3.8 million, despite only having around 1.5 million scheduled messages.

I investigated further, and found out that [Hangfire Console](https://github.com/pieceofsummer/Hangfire.Console) seems to be adding individual items starting with `console:` into the job-graph as well - one per line of logs!

I don't know who I think is the "sinner" here, but perhaps even storing in different collections could help?

For instance, instead of storing a `Key` of  `schedule:foo` in collection `jobGraph`, perhaps store it in a `jobGraph.schedule` collection.

That way, Hangfire Console would get into its own collection, and scheduled messages etc would as well.

It would speed up such queries heavily, since the queries always know where to get their data from, right?

**Comment by gottscj** [comment]
@ffMathy,

Could you try the query without Score? eg. only querying using _t and regex. you might want to put a limit on the result.
Its not possible to support atomicity if its not one collection. Hangfire was originally written with sql in mind. Im using the "BulkWrite" feature to ensure the needed transaction style api exposed by hangfire.

With these kinds of numbers and since you schedule them anyway, it might be better store the needed data and to poll using recurrent jobs.

I will see if I can help the queries by using the aggregation framework to force to match on _t first

Thanks!

**Comment by gottscj** [comment]
@ffMathy,

Please also try to remove the sort

Thanks!

**Comment by ffMathy** [comment]
@gottscj ah yes. Atomicity. But can't you then use Mongo Transactions, which was introduced in Mongo 4?

Here's without the sort, and without the `Score`:

![image](https://user-images.githubusercontent.com/2824010/112006146-bca55480-8b23-11eb-80d5-0cc2eb9ee8b7.png)

It looks like that's now using the `Key` index.

**Comment by ffMathy** [comment]
Also, generally, in my opinion, it should never take long to get the "next" item to schedule.

I'm not blaming your library here, but perhaps the overall way Hangfire was designed.

Taking the top job sorted ascending by schedule time, should not be an expensive operation, even with millions of records.

But I see how Hangfire doesn't exactly give you a lot of options here. Can we find a way to get around it somehow?

Quite curiously, just finding something based on score (and a large one even), takes "only" 3 seconds for 640,000 returned records:

![image](https://user-images.githubusercontent.com/2824010/112006826-5e2ca600-8b24-11eb-9b3c-f598d0ffafa2.png)

**Comment by gottscj** [comment]
@ffMathy,

Criticism is always welcome. without it I cant improve. :)

Thank you for the data. I will update you.

**Comment by gottscj** [comment]
@ffMathy,

quick question?

There is no "limit" operator used on the slow query you found in ATLAS? eg. ALL documents are returned?

**Comment by ffMathy** [comment]
Yes, looks like it. But it does have a "less than" and "greater than" on the `Score` field. I assumed that's how Hangfire handles fetching just some data, or something similar.

It does indeed seem a bit funny.

By the way, thank you so much for all of this. If you sign up for GitHub Sponsors, I'll sponsor you.

**Comment by ffMathy** [comment]
Interesting. Atlas just came up with this:

![image](https://user-images.githubusercontent.com/2824010/112127464-c2ec0d00-8bc5-11eb-9d06-36a5f5038d42.png)

And it's true - that particular index does not exist. I'll try creating it.

**Comment by ffMathy** [comment]
Note that it's also a composite index, like I suggested. While Mongo can merge individual indices, they are far from the same efficiency AFAIK.

**Comment by ffMathy** [comment]
Alright, that reduced cluster CPU usage from around 95% to 70% steadily.

However, the long queries remain. So there is an underlying architectural issue too.

**Comment by gottscj** [comment]
@ffMathy,

I have also tried to track down the offending query, and I cannot figure out where it comes from. that particular query should come with a limit of 1000 (hardcoded in Hangfire.Core). Can you confirm the query requests ALL documents for the given query, is it an update? if you manually dropped the collection, the indexes will not be rebuild again as they are created in a migration step

I also tried to change to use a compound unique key, but difference was minimal as mongo already uses the Score index.
a note on transactions, they are only supported if you have a replicaset, im not aware of the usage of Hangfire.Mongo, but I suspect its also being used with single instances, but I dont know. that could be easily added.

Could you check the indexes on the jobGraph, it should look something like this (split in two):
![image](https://user-images.githubusercontent.com/12970565/112128367-a7cdcd00-8bc6-11eb-8540-fb21addbf012.png)
![image](https://user-images.githubusercontent.com/12970565/112128478-c338d800-8bc6-11eb-825a-edde0ed2c71e.png)

**Comment by ffMathy** [comment]
They do indeed look like that. All those indices do indeed exist. Yet, creating the composite index that it suggested, worked just fine, and reduced CPU by around 15% for me, steadily.

I have never dropped the collection before, and I am 100% sure, because I am the only developer on this project.

**Comment by ffMathy** [comment]
Also, I looked closer into one of the offending queries that are still slow after the composite index. And indeed, you were right. The limit simply just isn't shown in the "Realtime" tab of queries.

Here is the full query:

```json
{
  "type": "command",
  "ns": "reshopper_prod.Hangfire__sirius_.jobGraph",
  "command": {
    "find": "Hangfire__sirius_.jobGraph",
    "filter": {
      "_t": "SetDto",
      "Key": {
        "$regex": "^schedule",
        "$options": ""
      },
      "Score": {
        "$gte": 0,
        "$lte": 1616493679
      }
    },
    "sort": {
      "Score": 1
    },
    "projection": {
      "Value": 1,
      "_id": 0
    },
    "limit": 1000,
    "$db": "reshopper_prod",
    "lsid": {
      "id": {
        "$uuid": "1ed1934e-a244-4c98-8716-9659bbcf4b36"
      }
    },
    "$clusterTime": {
      "clusterTime": {
        "$timestamp": {
          "t": 1616493679,
          "i": 187
        }
      },
      "signature": {
        "hash": {
          "$binary": "v9Hdt4LZyrxm+UrXwPWNkwe/2+E=",
          "$type": "00"
        },
        "keyId": {
          "$numberLong": "6904722023574929409"
        }
      }
    }
  },
  "planSummary": "IXSCAN { Score: 1 }",
  "keysExamined": 636915,
  "docsExamined": 636915,
  "fromMultiPlanner": true,
  "replanned": true,
  "replanReason": "cached plan was less efficient than expected: expected trial execution to take 24 works but it took at least 240 works",
  "cursorExhausted": true,
  "numYields": 2558,
  "nreturned": 5,
  "queryHash": "D8B51B3F",
  "planCacheKey": "5186805B",
  "reslen": 475,
  "locks": {
    "ReplicationStateTransition": {
      "acquireCount": {
        "w": 2559
      }
    },
    "Global": {
      "acquireCount": {
        "r": 2559
      }
    },
    "Database": {
      "acquireCount": {
        "r": 2559
      }
    },
    "Collection": {
      "acquireCount": {
        "r": 2559
      }
    },
    "Mutex": {
      "acquireCount": {
        "r": 1
      }
    }
  },
  "storage": {
    "data": {
      "bytesRead": 23982887,
      "timeReadingMicros": 24430
    }
  },
  "protocol": "op_msg",
  "durationMillis": 10629
}
```

**Comment by ffMathy** [comment]
What I find interesting is that it only returned 5 elements, had a limit of 1000, but also read 23 *megabytes* before returning those documents. It really suggests to me that the internal data structure has issues.

I also find it interesting that it was "replanned" for some reason. Not sure why that happened, but that can't be good for performance.

I understand your perspective on enabling transactions. That was not very thought-through of me. It would be a shame to require multi-node clusters indeed.

Any other ideas?

**Comment by gottscj** [comment]
@ffMathy,

Thanks, I see limit of 1000, that makes more sense. The SetDto's Scores are being updated by hangfire, but writes should not stop reads on mongo, so im not completely sure whats going on. The amount of documents thats needs to examined and sorted is quite high, and im not sure changing the schema would help that.

I think for now, your best option is to batch the required work into (alot) less workers and let the jobs run work in parallel.
Or if you can schedule the jobs at different times to split the workload.

Maybe schedule 1000 jobs which then starts 1500 Tasks in parallel.

I will need to study the schemas, and perhaps split them into separate collections. I like the idea of using transactions for systems that supports this.

**Comment by gottscj** [comment]
Also you could set the schedule poll interval in "BackgroundServerOptions", its default value is 15 seconds. depending on your schedule you could increase this value

**Comment by gottscj** [comment]
@ffMathy,

I just released 0.7.20. I added a field so the regex query is no longer used. I also added some optimizations when updating data.
Please let me know if it helps.

Thanks

**Comment by gottscj** [comment]
I also added the possibility to opt out of using a tailed capped collection to notify jobs added to queue, you should probably opt out as this is primarily used for fire-and-forget and it would be less strain on your db. just a suggestion

**Comment by ffMathy** [comment]
@gottscj thank you so much for your efforts here! I'll check it out Monday and report my findings. I'll also make good use of opting out of that feature. That should mean so much.

Cheers! 🍺 🙏

**Comment by gottscj** [comment]
Sure thing.

I hope it works out. Else we need to look at schemas and collections

**Comment by ffMathy** [comment]
Holy...!

That worked. CPU decreased from 80% stable, to varrying between 20% and 40% (still spikes once in a while, but that's due to other activity). In addition, there are no more slow Hangfire queries being reported - at all. They went from 15 seconds down to less than 200 milliseconds.

Thank you so much for this!

Oh, and we've been using this for several years. Perhaps bump to `1.0`? :smirk:

**Comment by gottscj** [comment]
great news!

if I bump to 1.0 I have no excuse :) but yes, its time.
maybe after next release where I will support transactions and $watch features.

![image](https://user-images.githubusercontent.com/12970565/112827418-9a18bb80-908e-11eb-8a99-21715877571b.png)

**Comment by ffMathy** [comment]
Is there any way that I can sponsor you? Are you on GitHub Sponsors?

**Comment by gottscj** [comment]
your very kind.
I'm not on sponsors, i did sign up though. 

thanks!

**Comment by ffMathy** [comment]
Alright. Let me know when you're signed up, and I'll become your first sponsor 🥂

**Comment by ffMathy** [comment]
By the way, here's the CPU utilization after the update:
![image](https://user-images.githubusercontent.com/2824010/112837456-339a9a00-909c-11eb-9c1a-41a65b7afdf1.png)

**Comment by ffMathy** [comment]
Hmmm, side note. All of our scheduled messages were cleared as part of the update 😦 

Maybe that was why performance improved. I'll let you know in a month or so, when we reach the same level of scheduled messages again.

**Comment by gottscj** [comment]
I needed to create a migration. If your migration strategy is drop,then it sounds about right. 😬
Yes please let me know! 👍

**Comment by ffMathy** [comment]
Oh my god, the strategy was drop for some reason indeed. Not sure why we chose that. We must have had issues with migrate in the past.

I think it's because we do a partial rollout on all of our horizontally-scaled nodes.

I think the regular migrate strategy caused issues sometimes if something has been deployed to node A, but not node B yet?

But I re-enabled it now. I'll let you know if that causes problems in the future :+1:

My bad! Sorry about that.

**Comment by gottscj** [comment]
@ffMathy,

Any chance you got some insights? no news is good news I hope?

**Comment by ffMathy** [comment]
@gottscj definitely. However, the load is still higher than we'd like. Not sure if that's due to this package though. I think that's just how Hangfire is built up.

We're also switching Hangfire over to a new PostgreSQL cluster soon either way, to offload the load from our main server.

**Comment by gottscj** [comment]
@ffMathy,

Thanks for the feedback. Good luck!
